addons/point_of_sale/controllers/main.py

```diff
@@ -84,3 +84,64 @@ def print_sale_details(self, date_start=False, date_stop=False, **kw):
         pdf, _ = request.env['ir.actions.report'].with_context(date_start=date_start, date_stop=date_stop)._render_qweb_pdf('point_of_sale.sale_details_report', r)
         pdfhttpheaders = [('Content-Type', 'application/pdf'), ('Content-Length', len(pdf))]
         return request.make_response(pdf, headers=pdfhttpheaders)
+
+    @http.route(['/pos/ticket'], type='http', auth="user", website=True)
+    def show_ticket_form(self, order_reference='', ticket_amount='', error_message=''):
+        return request.render("point_of_sale.ticket_entry", {
+            'order_reference': order_reference,
+            'ticket_amount': ticket_amount,
+            'error_message': error_message,
+        })
+
+    @http.route(['/pos/ticket/validate'], type='http', auth="user", website=True)
+    def show_ticket_validation_screen(self, order_reference='', ticket_amount=''):
+        if not order_reference or not ticket_amount:  # The route is called directly
+            return request.redirect('/pos/ticket')
+        partner = request.env.user.partner_id
+        template_vals = {
+            'partner': partner,
+            'format_amount': format_amount,
+            'env': request.env,
+            'order_reference': order_reference,
+            'ticket_amount': ticket_amount,
+            'address_url': f'/my/account?redirect=/pos/ticket/validate?order_reference={order_reference}%26ticket_amount={ticket_amount}',
+        }
+        # Check that an order exists for the information entered
+        pos_order = request.env['pos.order'].sudo().search([('amount_total', '=', ticket_amount), ('pos_reference', '=like', f'Order {order_reference}')])
```

Comment:
Thank you for the comments!

The issue is that the pos order isn't linked to a partner yet in this flow.
The order is placed from the front end (in store) => it doesn't necessarily have a customer set to it.

I understand the concern that it is not the safest way of doing it, but I can't see another one as we don't have any information on the order's " owner ".

The user is required to be logged in for all routes, and a sudo is needed otherwise the record rules are blocking the search.

I added a check on the search so that we only return orders if they have no partner or if the partner is linked to the current user, which is indeed better and should avoid risking displaying the invoice of other customers. (locally atm) 
Although if they are already invoiced it would try to show the /my/invoice which would redirect to the /my as the user wouldn't have access to it.

Still, I'm not sure how we could do more without ditching the manual route and only allowing to request the invoice from a QR code (in which we could add a token of some kind).

Please correct me if I'm wrong or if you have more ideas.

Reply:
We could probably implement a token system to identify orders?

Reply:
I definitely can, pos order just need to inherit the portal mixin to get all that ready to use.
And if we ditch the "manual" part of it (entering the reference/amount) and only use the QR code to access the routes, then we can just put the order id and access token in the QR/routes directly and bingo.

It's a valid solution, I will talk about it with CHKL tomorrow to see if we agree to do that. Too bad for the completely manual flow but can't help it.

The multi-company portal issue is what is really bothering me, I guess we will need to just return the PDF directly for now and maybe update later if we can improve the portal :frowning: